itests/api_v2_test.go

```diff
@@ -0,0 +1,291 @@
+
+	"github.com/filecoin-project/go-f3"
+	"github.com/filecoin-project/go-f3/certs"
+	"github.com/filecoin-project/go-f3/gpbft"
+	"github.com/filecoin-project/go-f3/manifest"
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/actors/policy"
+	"github.com/filecoin-project/lotus/chain/lf3"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/itests/kit"
+)
+
+func TestAPIV2_GetTipSetThroughRPC(t *testing.T) {
+	const (
+		timeout          = 2 * time.Minute
+		blockTime        = 10 * time.Millisecond
+		f3FinalizedEpoch = 123
+		targetHeight     = 20 + policy.ChainFinality
+	)
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	t.Cleanup(cancel)
+	kit.QuietMiningLogs()
+
+	mockF3 := newMockF3Backend()
+	subject, _, network := kit.EnsembleMinimal(t, kit.ThroughRPC(), kit.F3Backend(mockF3))
+	network.BeginMining(blockTime)
+	subject.WaitTillChain(ctx, kit.HeightAtLeast(targetHeight))
+
+	var (
+		heaviest = func(t *testing.T) *types.TipSet {
+			head, err := subject.ChainHead(ctx)
+			require.NoError(t, err)
+			return head
+		}
+		ecFinalized = func(t *testing.T) *types.TipSet {
+			head, err := subject.ChainHead(ctx)
+			require.NoError(t, err)
+			ecFinalized, err := subject.ChainGetTipSetByHeight(ctx, head.Height()-policy.ChainFinality, head.Key())
+			require.NoError(t, err)
+			return ecFinalized
+		}
+		tipSetAtHeight = func(height abi.ChainEpoch) func(t *testing.T) *types.TipSet {
+			return func(t *testing.T) *types.TipSet {
+				ts, err := subject.ChainGetTipSetByHeight(ctx, height, types.EmptyTSK)
+				require.NoError(t, err)
+				return ts
+			}
+		}
+		internalF3Error = errors.New("lost hearing in left eye")
+		plausibleCert   = func(t *testing.T) *certs.FinalityCertificate {
+			f3FinalisedTipSet := tipSetAtHeight(f3FinalizedEpoch)(t)
+			return &certs.FinalityCertificate{
+				ECChain: &gpbft.ECChain{
+					TipSets: []*gpbft.TipSet{{
+						Epoch: int64(f3FinalisedTipSet.Height()),
+						Key:   f3FinalisedTipSet.Key().Bytes(),
+					}},
+				},
+			}
+		}
+		implausibleCert = &certs.FinalityCertificate{
+			ECChain: &gpbft.ECChain{
+				TipSets: []*gpbft.TipSet{{
+					Epoch: int64(1413),
+					Key:   []byte(`🐠`),
+				}},
+			},
+		}
+	)
+
+	// The tests here use the raw JSON request form for testing to both test the API
+	// through RPC, and showcase what the raw request on the wire would look like at
+	// Layer 7 of ISO model.
+	for _, test := range []struct {
+		name       string
+		when       func(t *testing.T)
+		request    string
+		wantTipSet func(t *testing.T) *types.TipSet
+		wantErr    string
+	}{
+		{
+			name:       "no selector",
+			request:    `{"jsonrpc":"2.0","method":"Filecoin.ChainGetTipSet","id":1}`,
+			wantTipSet: heaviest,
+		},
+		{
+			name:       "latest tag",
+			request:    `{"jsonrpc":"2.0","method":"Filecoin.ChainGetTipSet","params":{"tag":"latest"},"id":1}`,
```

Comment:
```suggestion
			request:    ChainGetTipSetRequestBody(`{"tag":"latest"}`),
```

nit: factor out the boiler plate?

Reply:
I explicitly want the test to show the raw JSON-rpc request on the wire. See comment above the test table.